object_store/src/lib.rs

```diff
@@ -153,19 +153,22 @@ impl GoogleCloudStorage {
             }
         );
 
-        let location = location.to_string();
+        let location_string = location.to_string();
         let bucket_name = self.bucket_name.clone();
 
         let _ = tokio::task::spawn_blocking(move || {
             cloud_storage::Object::create(
                 &bucket_name,
                 &temporary_non_streaming,
-                &location,
+                &location_string,
                 "application/octet-stream",
             )
         })
         .await
-        .context(UnableToPutDataToGcs)?;
+        .context(UnableToPutDataToGcs {
+            bucket: self.bucket_name.clone(),
```

Comment:
I did this too recently! @carols10cents helpfully explained that the `context()` calls `into()` on the argument inside a closure when the error is triggered - this lets us avoid cloning the string in the happy path so we only pay the price for it if the error occurs which is great!

Rust just needs a little help figuring out that it can do this with a `str`:

```suggestion
            bucket: &*self.bucket_name,
```

This [derefs](https://doc.rust-lang.org/std/string/struct.String.html#deref-methods) the `String` to a `str` and then takes a reference to the resulting `str` so we wind up passing in a `&str` (I believe anyway). Since there is a `From<&str> for String` impl, this has an `into()` string for the closure to call :)

This lets us answer have our cake and eat it:

> Should bucket and location represent String objects or &str?

The error can take a String so it owns the values (no lifetimes for errors is nicer) but we don't have to clone everywhere!

We can probably apply this to all the error paths here :)

Reply:
Yes, this is something that I find confusing about the use of Snafu's `context` calls as well -- they are always constructed on the happy, but they don't have the same type as the erorr variant (so in this case the `bucket` field in `Context(UnableToPutDataToGcs { ...})`  is not a `String` but rather `impl Into<String>`

Reply:
> Rust just needs a little help figuring out that it can do this with a `str`:

Kind of. The problem with `bucket: self.bucket_name` is that it would try to move ownership of `bucket_name` but that's disallowed because you don't own `self` (it's `&self`).

You don't have to deref-and-re-reference `String` either, as `&String` also implements `Into<String>`:

```rust
fn x(_: impl Into<String>) {}

fn main() {
    let s = String::new();
    x(&s);
}
```

So should be able to be:

```rust
bucket: &self.bucket_name,
```

Reply:
Applied the `bucket: &self.bucket_name` change. Appreciate the explanation.